### PR TITLE
intel_gpu: sdpa: int8 K in the micro-kernel SDPA path

### PR DESCRIPTION
**Pull request by dmenig**
Extends the micro-kernel SDPA generator and kernel so attention can consume an `i8` K, and apply
a per-key mask without materialising it as a full score tile.

The integer K·Q path is a memory-traffic reduction rather than an arithmetic one, which is what
makes it worth doing: in attention with a long key sequence K is among the largest tensors the
kernel touches, and the kernel is bandwidth-bound on it well before it is compute-bound.

- **`i8` K for the K·Q product** (`I8_KQ`): s8 × s8 → s32 on the systolic pipe. Q is narrowed
  to s8 in-kernel while it is staged to SLM (it arrives from the graph already as integer-valued
  codes; see the requirements below), so K is never dequantized and its load traffic is
  halved relative to f16.
- **Per-key mask** (`MASK_PER_KEY`): a static-shape `[b, h, 1, k]` mask carries one value per
  key, so it is loaded as a coalesced vector and broadcast along the query direction instead of
  being materialised as a full S-shaped tile through a transposed load. A bf16 mask arrives
  through the same `half*` argument, so its 16-bit values are reinterpreted as bf16 and widened to
  float, as the general path does.

### The integer V·S path is no longer in this PR

An earlier revision of this PR also quantized the softmax numerator and ran V·S on the integer
pipe (`I8_VS`). **It was wrong and it has moved to #37859.** The V·S gemm reads V as its A
operand on the systolic pipe and contracts it over the keys, so it needs the key axis
contiguous; `problem_vs.A.layout` is `MatrixLayout::N` (head-size contiguous), which is right
for f16, where gemmstone repacks, and wrong for s8. On an Arc iGPU the consequences were: at head
64 no s8 V·S microkernel matched and the implementation then abandoned micro SDPA entirely and
fell back to `sdpa_opt` — so an `i8` V also cost the working `i8` K path — and at head 128 the
microkernel was selected and returned cosine 0.004 against the reference. Cosine is
scale-invariant, so that is not a missing dequantization factor.

`I8_KQ` is unaffected because K's A operand is already `MatrixLayout::T`. The legal layout for
V is the transposed one, which is supplied by the `TRANSPOSE_V` plumbing in #37859, so `I8_VS`
lands there gated on it (`micro_i8_vs()` requires `micro_transpose_v()`, with a matching
`#error` in the kernel) and with tests covering both value layouts.

### Where the integer path does and does not engage

`I8_KQ` engages only when K arrives from the graph **already quantized** to `i8`. The plugin's
compressed KV cache is `i8` as well, but it carries scales and zero-points and must keep its
decompression path, so `I8_KQ` is gated on `!is_paged_attention && !is_kv_compressed` in both
the JIT constants and the microkernel selection, and the kernel refuses the combination with an
`#error`.

### What the integer path requires of the graph

It carries no scale of its own; the graph that produces the `i8` K owns the dequantization, and
the matcher in #37859 (the follow-up PR that routes a quantized K into micro-kernel SDPA) is
what sets this up.

Q must arrive as integer-valued f16 codes (already divided by its own quantization step), so the
in-kernel rounding to s8 is exact, and the product of K's and Q's steps must be folded into the
SDPA `scale` input, which the kernel applies to S. The generator selects `I8_KQ` only for an f16
Q, a head size that is a multiple of 4 (four head-dim values pack into one dword) and a `D_MAX/4`
that is a multiple of the subgroup size (one packed row per lane); see `micro_i8_kq_shape_ok`.
Where any of those does not hold, the gemm dequantizes the `i8` K instead and the result is
unchanged — covered by a test below.

### Measurement

No end-to-end speedup is measured here. The traffic argument above is structural and checkable
by inspection; the actual gain depends on how bandwidth-bound the target attention shape is, and
nothing in-tree exercises the integer path until #37859 lands (see below). I will run whatever
public quantized models you consider representative.

### Requirements and caveats

This depends on #37856 — the core SDPA op must accept an `i8` K next to an f16 Q; without it the
graph fails with `Mixed input types are not supported`. This branch is based on it, so its
commits show up in this PR's history until it merges.

**Merged on its own, `I8_KQ` is never selected.** Nothing in-tree routes a quantized K to
micro-kernel SDPA; the matcher that does, and the op/primitive plumbing it needs, live in #37859.
That makes this PR inert rather than wrong for quantized operands if it lands first.

**`MASK_PER_KEY` is the one path that changes the kernel generated for existing f16/bf16
models**, and only when the mask has a static `[b, h, 1, k]` shape. A bf16 per-key mask case is
added to `sdpa_gpu_test`; the existing f16 cases with unequal query/key lengths already use a
mask of that shape and so cover the f16 arm.

The per-key mask shape is only inspected once its rank is known. `sdpa_has_runtime_attn_mask_input`
admits a dynamic-rank mask and `PartialShape::size()` asserts on one, which would have failed
micro-kernel JIT generation for a model that should simply have taken the general mask path. I
could not construct a reproducer — every route into the generator canonicalizes to rank 4 first —
so this is a closed latent inconsistency rather than a demonstrated crash, and it has no test.

### Testing

`sdpa_gpu_test` gains `sdpa_micro_i8`, which constructs the GPU primitive directly and so can
exercise an `i8` K without waiting for #37859's matcher. Each case runs one set of
integer-valued inputs twice — an `i8` K on `sdpa_micro`, and those same codes held as f16 on
`sdpa_ref` — and compares. Because the caller's contract folds the dequantization scale into the
op's `scale` input, the f16 run over the same codes is the correct answer, so the integer path
must reproduce it rather than approximate it. Every case also asserts which implementation
actually ran, so a case that stops reaching the path it was written for fails instead of passing
vacuously. Four cases, all at cosine ≥ 0.999 on an Arc iGPU (Xe-LPG): the integer contraction at
head 64 and head 128, the head-32 shape that makes `I8_KQ` decline and fall back to a
dequantizing gemm, and a single-token shape that does not use micro SDPA at all.

f16/bf16 results are unchanged: `--gtest_filter='*sdpa*:*rope*:*paged_attention*'` on the same
Arc iGPU gives **327 run, 283 passed, 44 skipped, 0 failed** (RoPE is included because #37859
will fuse it into this kernel).

The integer path has additionally been run end-to-end through #37859's matcher on the same
device, on an in-house int8 quantization-aware-trained video transformer (ViT-B class, head size
64) checked against its own f16 outputs; that validation ships with #37859.

The micro-kernel SDPA generator targets Xe-HPG/Xe-LPG. Xe-LP (Gen12 Iris Xe) is unverified; the
generator is not expected to be selected there, which leaves this path inert.

### CODE REVIEW

File: docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/sequence/scaled-dot-product-attention.rst [hunk 1]
```diff
@@ -74,6 +74,14 @@ omitting training-related parameter.
 
     The sink input is available since 2025.4 OpenVINO release.
 
+.. note::
+
+    A *T_QUANT* ``key`` or ``value`` is available since 2026.4 OpenVINO release. The pseudo-code
+    above describes the operation for a *T* ``key`` and ``value`` only. When either is *T_QUANT*
+    it holds quantization codes rather than values, and the mapping from codes to values is
+    defined by the implementation that consumes the operation, not by this specification. An
+    implementation that does not define such a mapping is expected to reject the operation.
```
[the remark this answers is not in the document]
Comment (dmenig) [reply docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/sequence/scaled-dot-product-attention.rst:83]: This file and this line are `#37856`'s; they are in this diff only because this PR is stacked on it. The same finding was raised there and answered there on 2026-09-15 — https://github.com/openvinotoolkit/openvino/pull/37856#discussion_r4016655481 — and the answer has not changed, so briefly:

Scale and zero-point operands have nowhere to go on this op. Its optional inputs are positional — attention_mask at 3, scale at 4, sink at 5 — so a key scale would have to sit at 6 and beyond, and no model could supply one without also supplying a sink it does not have. Giving those operands a defined code-to-value mapping would also make quantized attention a semantic every implementation owes, the reference and template backends included, which is a much larger change to a released opset than the type relaxation it would replace.

The self-describing form already exists one level down, which is where it belongs: `KVCacheCompression` in the GPU plugin matches the internal `IndirectSDPA` and rebuilds it carrying compression scales and zero points as real operands. An implementation that can honour a mapping states it there; the public op deliberately does not.

What `#37856` does instead is make the ambiguity impossible to hit silently: the specification says an implementation without a code-to-value mapping is expected to reject the op, all seven in-tree consumers decline or report the node by name rather than reading codes at the query's precision, and a serialization round-trip test pins that a u4 key and i8 value survive `Serialize` with their types intact. The residual exposure is an out-of-tree consumer that ignores the spec text, and that is a trade being made knowingly rather than overlooked.

Settling this is a maintainer's call, and the `#37856` body asks for it explicitly. Please take it there rather than here.

File: src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp [hunk 2]
```diff
@@ -502,6 +502,18 @@ ov::pass::StateManagementPattern::StateManagementPattern(PaParams& pa_params,
     auto sdpa_variants = std::make_shared<Or>(OutputVector{sdpa_with_4_inputs, sdpa_with_5_inputs, sdpa_with_6_inputs});
 
     ov::matcher_pass_callback callback = [=, &pa_params, &results, &var_ids_to_remove](Matcher& m) {
+        // PagedAttentionExtension has no operand able to carry the mapping from quantization
+        // codes to values. SDPAToPagedAttention is an explicit conversion request, so report it
+        // against the node rather than leaving a half-converted model behind.
+        const auto sdpa_root = ov::as_type_ptr<v13::ScaledDotProductAttention>(m.get_match_root());
+        NODE_VALIDATION_CHECK(sdpa_root.get(),
+                              !v13::ScaledDotProductAttention::has_quantized_kv(*sdpa_root),
```
[the remark this answers is not in the document]
Comment (dmenig) [reply src/common/transformations/src/transformations/paged_attention/state_management_pattern.cpp]: Agreed, and this is already fixed — but you are right that it was not fixed *in this diff*, and that is worth stating plainly rather than waving at another PR.

The pre-scan landed on `#37856` as `b3b77e0628` on 2026-09-17: the check is gone from the matcher callback and now runs at the top of `SDPAToPagedAttention::run_on_model`, beside the assertions that already run there before anything is mutated. It carries two mutation-checked regression tests, `SDPAToPAQuantizedKVTest.RejectedBeforeAnyLayerIsConverted` (two stacked OPT-125m attention layers, the second with an i8 K/V cache — it fails against the old in-callback guard with one `PagedAttentionExtension` built and one SDPA surviving, exactly the half-converted model you describe) and `StateConcatSDPAQuantizedKVShouldNotFuse`.

This PR's base was a replay of `#37856`'s *pre-fix* commit, so the code you are looking at genuinely was the unfixed version when you reviewed it. The base has been re-cut on top of `#37856`'s current content, so at the new head this file matches `#37856` and the finding is gone from this diff too.

File: src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa/sdpa_gen_micro.cpp [hunk 3]
```diff
@@ -1442,6 +1469,17 @@ JitConstants SDPAMicroGenerator::get_jit_constants(const kernel_impl_params& par
...
+
+        // a [b, h, 1, k] mask carries one value per KEY and none per query, yet the
+        // general path in sdpa_micro.cl materialises it as a full S-shaped tile through a
+        // transposed load -- sg_tile_n times the traffic and registers for the same
+        // information, plus a half->float tile copy and a full-tile add. MASK_PER_KEY loads a
+        // coalesced sg_tile_m vector instead and broadcasts it along the query direction with
+        // the machinery REMAINDER_K's k_mask already uses.
+        const auto& msk_shape = params.input_layouts[3].get_partial_shape();
+        bool per_key =
+            msk_shape.size() == 4 && msk_shape[2].is_static() && msk_shape[2].get_length() == 1 && msk_shape[3].is_static() && msk_shape[3].get_length() > 1;
```
[the remark this answers is not in the document]
Comment (dmenig) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa/sdpa_gen_micro.cpp]: Correct, and fixed. `PartialShape::size()` is an `OPENVINO_ASSERT`, not a debug-only check, and `sdpa_has_runtime_attn_mask_input()` does explicitly admit a dynamic-rank mask — it returns true on `!rank().is_static()`. So the two were inconsistent and this code could throw out of JIT generation for a model that should simply have taken the general mask path.

The rank is now established before the shape is indexed, and a mask whose rank is unknown leaves `MASK_PER_KEY` off:

```cpp
const bool per_key = msk_shape.rank().is_static() && msk_shape.size() == 4 && msk_shape[2].is_static() &&
                     msk_shape[2].get_length() == 1 && msk_shape[3].is_static() && msk_shape[3].get_length() > 1;
```

One caveat I would rather state than leave implied: I could not build an end-to-end reproducer. Every route I tried into this generator hands it a rank-4 layout, because the plugin canonicalizes earlier, so I have no failing case to point at and no regression test for it. What I can say is that the assert is real, the helper this code sits behind is written to accept the input that trips it, and the guard costs nothing. I am treating it as a latent inconsistency worth closing rather than a demonstrated crash.

File: src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa/sdpa_gen_micro.cpp [hunk 4]
```diff
@@ -1228,6 +1236,25 @@ JitConstants SDPAMicroGenerator::get_jit_constants(const kernel_impl_params& par
...
+    // scale input, so the in-kernel rounding that packs Q into SLM is exact.
+    // An i8 K or V here is an externally quantized operand, not the plugin's compressed KV
+    // cache: that one is i8 as well but carries scales/zero-points and must keep taking the
+    // decompression paths above, so it never selects these branches.
+    const bool external_i8_kv = !config.is_paged_attention && !config.is_kv_compressed;
+    const bool i8_kq = external_i8_kv && K.data_type == ov::element::i8 && micro_i8_kq_shape_ok(Q, k_head_size, d_max, get_subgroup_size(device_info.arch));
+    jit.make("I8_KQ", i8_kq ? 1 : 0);
+
+    const bool i8_vs = external_i8_kv && V.data_type == ov::element::i8;
+    jit.make("I8_VS", i8_vs ? 1 : 0);
```
[the remark this answers is not in the document]
Comment (dmenig) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa/sdpa_gen_micro.cpp]: You were right, and the tests found something worse than the gap you were pointing at. Thank you for pushing on this one.

**What the tests are.** `sdpa_gpu_test` gains `sdpa_micro_i8`, built the way you suggested — the GPU primitive directly, no graph matcher needed. Each case runs the same integer-valued inputs twice, once with an i8 key on `sdpa_micro` and once with those same codes held as f16 on `sdpa_ref`, and compares. Because the caller's contract is that the dequantization scale is folded into the op's scale input, the f16 run over the same codes *is* the right answer, so the integer path has to reproduce it rather than approximate it. Every case also asserts which implementation actually ran, so a case that stops reaching the path it was written for fails instead of passing quietly. Four cases: the integer contraction at head 64 and head 128, the head-32 shape that makes `I8_KQ` decline and fall back to a dequantizing gemm, and a single-token shape that does not use the micro kernel at all. All four pass at cosine ≥ 0.999.

**What they found.** The integer `V*S` branch was broken, and in two different ways. At head 64 the s8 `V*S` microkernel does not exist — oneDNN answers "No matching kernel" — and the implementation then abandons the micro kernel *entirely* and silently falls back to `sdpa_opt`, so an i8 value cost you the working i8-key path as well. At head 128 the microkernel does get built, and the answer is wrong: cosine 0.004 against the reference. Cosine is scale-invariant, so that is not a missing dequantization factor.

**Root cause.** The `V*S` gemm reads V as its A operand on the systolic pipe and contracts it over the keys, so it needs the key axis contiguous. `problem_vs.A.layout` is `MatrixLayout::N` — head-size contiguous — which is right for f16, where gemmstone repacks, and wrong for s8. `I8_KQ` works precisely because K's A operand is already `MatrixLayout::T`. I confirmed this on an Arc iGPU rather than inferring it: with a V that is symmetric in (key, head-dim), so that both layouts read the same matrix, `N` reproduces the failure and `T` matches the reference exactly, value for value. The intermediate diagnostic is worth recording too — with V held constant, the first eight head-dim values of every sixteen were correct and the next eight were not, and with V varying only along the head dim the output came out as `(V[d] + V[d+1]) / 2`, the signature of adjacent head-dim values being folded together as if they were the packed contraction axis.

**What changed.** The integer `V*S` path is **removed from this PR**. It is only correct on a value tensor laid out with the key axis contiguous, which needs the transposed-V plumbing that lives in `#37859`; there is no way to make it correct here, and gating it off would just leave unreachable code upstream. It moves to `#37859`, where it is gated on that layout — `micro_i8_vs()` requires `micro_transpose_v()`, with a matching `#error` in the kernel — and carries its own test suite covering both layouts, including the ordinary layout as a regression test for exactly this bug.

This PR is therefore the i8 key and the per-key mask, both tested. Full GPU run on an Arc iGPU, `*sdpa*:*rope*:*paged_attention*`: 327 tests, 283 passed, 44 skipped, 0 failed.

**Comment by p-durandin** [comment]
build_jenkins
